helix-term/src/ui/editor.rs

```diff
@@ -437,51 +421,107 @@ impl EditorView {
+            theme: &Theme,
+            config: &Config,
+            is_focused: bool,
+            width: usize,
+        ) -> Box<dyn Fn(usize, bool) -> Option<(String, Style)>> {
+            let text = doc.text().slice(..);
+            let last_line = view.last_line(doc);
+            // Whether to draw the line number for the last line of the
+            // document or not.  We only draw it if it's not an empty line.
+            let draw_last = text.line_to_byte(last_line) < text.len_bytes();
 
-            let text = if line == last_line && !draw_last {
-                "    ~".into()
-            } else {
-                let line = match config.line_number {
-                    LineNumber::Absolute => line + 1,
-                    LineNumber::Relative => {
-                        if current_line == line {
-                            line + 1
-                        } else {
-                            abs_diff(current_line, line)
-                        }
-                    }
-                };
-                format!("{:>5}", line)
-            };
-            surface.set_stringn(
-                viewport.x + 1,
-                viewport.y + i as u16,
-                text,
-                5,
-                if selected && is_focused {
-                    linenr_select
+            let linenr = theme.get("ui.linenr");
+            let linenr_select: Style = theme.try_get("ui.linenr.selected").unwrap_or(linenr);
+
+            let current_line = doc
+                .text()
+                .char_to_line(doc.selection(view.id).primary().cursor(text));
+
+            let config = config.line_number;
+
+            Box::new(move |line: usize, selected: bool| {
+                if line == last_line && !draw_last {
+                    Some(("    ~".into(), linenr))
                 } else {
-                    linenr
-                },
-            );
+                    let line = match config {
+                        LineNumber::Absolute => line + 1,
+                        LineNumber::Relative => {
+                            if current_line == line {
+                                line + 1
+                            } else {
+                                abs_diff(current_line, line)
+                            }
+                        }
+                    };
+                    let style = if selected && is_focused {
+                        linenr_select
+                    } else {
+                        linenr
+                    };
+                    Some((format!("{:>1$}", line, width), style))
+                }
+            })
+        }
+
+        let gutters: &[(
+            fn(
+                &Document,
+                &View,
+                &Theme,
+                &Config,
+                bool,
+                usize,
+            ) -> Box<dyn Fn(usize, bool) -> Option<(String, Style)>>,
+            usize,
+        )] = &[(diagnostic, 1), (line_number, 5)];
```

Comment:
Removing a gutter item (`line_number` for example) still leaves unused extra spaces, since the gutter width is hard coded...

https://github.com/helix-editor/helix/blob/f9375f449c8b7bbf792c88b89903fe38d524f2e5/helix-view/src/view.rs#L83-L87

... and then used for rendering text:

https://github.com/helix-editor/helix/blob/79e6a010021b8a446fdb80f474839ed024ee00a3/helix-term/src/ui/editor.rs#L75-L90

I suppose we would need to store the active gutter items with default widths in `View` and then use it as necessary. Maybe a separate `Gutter` component that handles it's own rendering ?

Reply:
Yeah I didn't make the change yet. I was thinking `inner_area` could consume a `gutter_width: u16`?

Now that I think about it, storing the gutters on the View might be nice since we could enable/disable these per view.

Reply:
A `gutter_width` param for `inner_area` would probably add friction to it's usage (it's used in most of the scrolling commands). Storing the gutters in view also has the added advantage you mentioned.